File: rust/integration-gateway/src/integrations/repository.rs
use serde_json::Value;
use sqlx::{PgPool, Row};

use super::model::IntegrationRow;

/// Sole `errors` sentinel, matching Django's `ERROR_TOKEN_REFRESH_FAILED` (posthog/models/integration.py).
pub const ERROR_TOKEN_REFRESH_FAILED: &str = "TOKEN_REFRESH_FAILED";

/// Fetch integration rows by id. Read-only (SELECT). `config`/`sensitive_config` are `jsonb`
/// and decode straight into `serde_json::Value` (sqlx `json` feature).
///
/// `id`/`team_id` are cast to `bigint` so decoding as `i64` is reliable whether the underlying
/// columns are `int4` (Django `AutoField`) or `int8` (`BigAutoField`).
pub async fn fetch_by_ids(pool: &PgPool, ids: &[i64]) -> Result<Vec<IntegrationRow>, sqlx::Error> {
    if ids.is_empty() {
        return Ok(Vec::new());
    }

    let rows = sqlx::query(
        "SELECT id::bigint AS id, team_id::bigint AS team_id, kind, config, sensitive_config \
         FROM posthog_integration WHERE id = ANY($1)",
    )
    .bind(ids)
    .fetch_all(pool)
    .await?;

    let mut out = Vec::with_capacity(rows.len());
    for row in rows {
        out.push(IntegrationRow {
            id: row.try_get("id")?,
            team_id: row.try_get("team_id")?,
            kind: row.try_get("kind")?,
            config: row.try_get("config")?,
            sensitive_config: row.try_get("sensitive_config")?,
        });
    }
    Ok(out)
}

/// Fetch a single row by id (used to re-read under the refresh lock so we don't refresh a token a
/// concurrent head already rotated). Returns `None` if the row no longer exists.
pub async fn fetch_one(pool: &PgPool, id: i64) -> Result<Option<IntegrationRow>, sqlx::Error> {
    Ok(fetch_by_ids(pool, &[id]).await?.into_iter().next())
}

/// Persist a successful refresh: new plaintext `config` and re-encrypted `sensitive_config`, and
/// clear any prior refresh error. Mirrors Django's `refresh_access_token` write.
pub async fn update_after_refresh(
    pool: &PgPool,
    id: i64,
    config: &Value,
    sensitive_config: &Value,
) -> Result<(), sqlx::Error> {
    sqlx::query(
        "UPDATE posthog_integration SET config = $1, sensitive_config = $2, errors = '' WHERE id = $3",
    )
    .bind(config)
    .bind(sensitive_config)
    .bind(id)
    .execute(pool)
    .await?;
    Ok(())
}

/// Record a failed refresh so the app surfaces "reconnect this integration" (same sentinel Django
/// sets). Leaves the stored tokens untouched.
pub async fn mark_refresh_failed(pool: &PgPool, id: i64) -> Result<(), sqlx::Error> {
    sqlx::query("UPDATE posthog_integration SET errors = $1 WHERE id = $2")
        .bind(ERROR_TOKEN_REFRESH_FAILED)
        .bind(id)
        .execute(pool)
        .await?;
    Ok(())
}

File: nodejs/src/integration-gateway/repository.ts
import { PostgresRouter, PostgresUse } from '~/common/utils/db/postgres'

import { IntegrationRow } from './types'

/** Sole `errors` sentinel, matching Django's `ERROR_TOKEN_REFRESH_FAILED` (posthog/models/integration.py). */
export const ERROR_TOKEN_REFRESH_FAILED = 'TOKEN_REFRESH_FAILED'

/**
 * Data access for `posthog_integration`. `config`/`sensitive_config` are `jsonb` and decode
 * straight into objects; `id`/`team_id` are `int4` (Django `AutoField`) so decode as numbers —
 * matching `IntegrationManagerService`'s existing query (no `::bigint` cast, which would make
 * node-postgres return a string).
 */
export class IntegrationRepository {
    constructor(private postgres: PostgresRouter) {}

    async fetchByIds(ids: number[]): Promise<IntegrationRow[]> {
        if (ids.length === 0) {
            return []
        }
        const response = await this.postgres.query<IntegrationRow>(
            PostgresUse.COMMON_READ,
            `SELECT id, team_id, kind, config, sensitive_config FROM posthog_integration WHERE id = ANY($1)`,
            [ids],
            'integrationGatewayFetchByIds'
        )
        return response.rows
    }

    /**
     * Re-read a single row under the refresh lock, from the PRIMARY (not the replica): a stale
     * replica read would defeat the point of re-reading under the lock and could refresh with an
     * already-rotated refresh token, revoking the grant on strict-rotation providers. Returns null
     * if the row no longer exists.
     */
    async fetchOneForUpdate(id: number): Promise<IntegrationRow | null> {
        const response = await this.postgres.query<IntegrationRow>(
            PostgresUse.COMMON_WRITE,
            `SELECT id, team_id, kind, config, sensitive_config FROM posthog_integration WHERE id = $1`,
            [id],
            'integrationGatewayFetchOneForUpdate'
        )
        return response.rows[0] ?? null
    }

    /**
     * Persist a successful refresh and clear any prior refresh error, guarded by the encrypted
     * `refresh_token` we read under the lock (compare-and-swap). The Redis lock only excludes other
     * gateway heads; a Django reconnect can still write concurrently. Guarding on the exact stored
     * ciphertext means if anything changed the row since we read it (a reconnect rotating the token),
     * the update matches 0 rows and we discard rather than clobber the new credentials. Returns true
     * iff the row was updated.
     */
    async updateAfterRefresh(
        id: number,
        config: Record<string, any>,
        sensitiveConfig: Record<string, any>,
        expectedRefreshToken: string
    ): Promise<boolean> {
        const response = await this.postgres.query(
            PostgresUse.COMMON_WRITE,
            `UPDATE posthog_integration SET config = $1, sensitive_config = $2, errors = ''
             WHERE id = $3 AND sensitive_config->>'refresh_token' = $4`,
            [JSON.stringify(config), JSON.stringify(sensitiveConfig), id, expectedRefreshToken],
            'integrationGatewayUpdateAfterRefresh'
        )
        return (response.rowCount ?? 0) > 0
    }

    /**
     * Record a failed refresh: persist the updated backoff `config` and set the `TOKEN_REFRESH_FAILED`
     * sentinel (same one Django sets, surfaced as "reconnect this integration"). Compare-and-swap
     * guarded on the encrypted `refresh_token` read under the lock, so a concurrent reconnect — which
     * resets the backoff and clears errors — is never clobbered by our stale failure state. Returns
     * true iff the row was updated.
     */
    async recordRefreshFailure(
        id: number,
        config: Record<string, any>,
        expectedRefreshToken: string
    ): Promise<boolean> {
        const response = await this.postgres.query(
            PostgresUse.COMMON_WRITE,
            `UPDATE posthog_integration SET config = $1, errors = $2
             WHERE id = $3 AND sensitive_config->>'refresh_token' = $4`,
            [JSON.stringify(config), ERROR_TOKEN_REFRESH_FAILED, id, expectedRefreshToken],
            'integrationGatewayRecordRefreshFailure'
        )
        return (response.rowCount ?? 0) > 0
    }
}
